### PR TITLE
Convert C++ mutation classes to polymorphic value types

### PR DESCRIPTION
**Pull request by wilhuff**
Note:
  * There remain some parts of these that are unimplemented; filling those out will will come separately.
  * Callers still use e.g. `std::unique_ptr<Mutation>`, mostly tests. This is not harmful because this code isn't yet used in any production capacity. I'll address it separately.

The general pattern here was laid down in #3529, and all the general commentary about how this works still applies. My take is that this approach is about as close as we're going to get to having polymorphic C++ "just work":
  * You can add these things to collections and `operator==` does the right thing.
  * You can accept a parameter of `const Mutation&` and subclasses are accepted
  * You can accept a parameter of `Mutation` and subclasses are also accepted without slicing
  * Subclass types can be recovered with a casting constructor, similar to the way we cast on the other platforms.
  * Far less e.g. `std::shared_ptr<const Mutation>` littering every parameter and variable declaration we have.

The cost is:
  * Base classes have to duplicate their external interface: once as virtual methods on the protected inner class and again as non-virtual methods that delegate
  * Subclasses have to implement a few extra constructors
  * Subclasses are mostly implemented in the `Rep` rather than where you might logically think they go using conventional polymorphism

### CODE REVIEW

**Comment by wilhuff** [comment]
Note the diffbase on #3551.

File: Firestore/core/src/firebase/firestore/model/set_mutation.h [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    size_t Hash() const override;
+
+    std::string ToString() const override;
+
+   private:
+    ObjectValue value_;
+  };
+
+  const Rep& set_rep() const {
+    return static_cast<const Rep&>(rep());
```
Comment (mikelehen) [inline Firestore/core/src/firebase/firestore/model/set_mutation.h]: I wonder if it's worth adding an assert on type() == Type::Set here?
(similar for patch_rep())

File: Firestore/core/src/firebase/firestore/model/delete_mutation.h [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    MaybeDocument ApplyToRemoteDocument(
+        const absl::optional<MaybeDocument>& maybe_doc,
+        const MutationResult& mutation_result) const override;
+
+    absl::optional<MaybeDocument> ApplyToLocalView(
+        const absl::optional<MaybeDocument>& maybe_doc,
+        const absl::optional<MaybeDocument>&,
+        const Timestamp&) const override;
+
+    // Does not override Equals; Mutation's version is sufficient.
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/model/delete_mutation.h]: Done.

File: Firestore/core/src/firebase/firestore/model/delete_mutation.h [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+ public:
+  DeleteMutation(DocumentKey key, Precondition precondition);
+
+  /**
+   * Casts a Mutation to a DeleteMutation. This is a checked operation that will
+   * assert if the type of the Mutation isn't actually Type::Delete.
+   */
+  explicit DeleteMutation(const Mutation& mutation);
+
+  /** Creates an invalid DeleteMutation instances. */
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/model/delete_mutation.h]: Done.

File: Firestore/core/src/firebase/firestore/model/delete_mutation.h [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+namespace model {
+
+/** Represents a Delete operation. */
+class DeleteMutation : public Mutation {
+ public:
+  DeleteMutation(DocumentKey key, Precondition precondition);
+
+  /**
+   * Casts a Mutation to a DeleteMutation. This is a checked operation that will
+   * assert if the type of the Mutation isn't actually Type::Delete.
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/model/delete_mutation.h]: It's ... uh ... right there, in `delete_mutation.cc`, line 37 😬.

(That is to say, great catch!)

File: Firestore/core/src/firebase/firestore/model/patch_mutation.h [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+                FieldMask mask,
+                Precondition precondition);
+
+  /**
+   * Casts a Mutation to a PatchMutation. This is a checked operation that will
+   * assert if the type of the Mutation isn't actually Type::Patch.
+   */
+  explicit PatchMutation(const Mutation& mutation);
+
+  /** Creates an invalid PatchMutation instances. */
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/model/patch_mutation.h]: Done.

File: Firestore/core/src/firebase/firestore/model/set_mutation.h [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+ public:
+  SetMutation(DocumentKey key, ObjectValue value, Precondition precondition);
+
+  /**
+   * Casts a Mutation to a SetMutation. This is a checked operation that will
+   * assert if the type of the Mutation isn't actually Type::Set.
+   */
+  explicit SetMutation(const Mutation& mutation);
+
+  /** Creates an invalid SetMutation instances. */
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/model/set_mutation.h]: Done.

File: Firestore/core/src/firebase/firestore/model/set_mutation.h (hunk 1 above)
Comment (mikelehen) [reply Firestore/core/src/firebase/firestore/model/set_mutation.h]: To be fair, when I made this comment, it was not explicitly checked in the casting constructor, so this would have addressed a real failure mode. 😛

File: Firestore/core/src/firebase/firestore/model/mutation.h [hunk 7]
```diff
@@ -199,149 +214,82 @@ class Mutation {
...
+      const Timestamp& local_write_time) const {
+    return rep_->ApplyToLocalView(maybe_doc, base_doc, local_write_time);
+  }
 
   friend bool operator==(const Mutation& lhs, const Mutation& rhs);
 
- protected:
-  Mutation(DocumentKey&& key, Precondition&& precondition);
+  size_t Hash() const {
+    return rep_->Hash();
```
[the remark this answers is not in the document]
Comment (mikelehen) [reply Firestore/core/src/firebase/firestore/model/mutation.h]: Hrm...  Is it true that dereferencing nullptr is undefined behavior? https://stackoverflow.com/a/28573259

If so and we can't statically prevent calling methods on invalid instances, it seems dangerous not to have explicit checks.  It seems entirely plausible for a coding mistake to result in accidentally calling a method on an invalid instance (e.g. because it was moved-from in another part of the code).

I hoped dereferencing a null shared_ptr<> would fail in a deterministic sane way, but based on https://stackoverflow.com/questions/34409299/why-doesnt-stdshared-ptr-dereference-throw-a-null-pointer-exception-or-simil it sounds like it intentionally doesn't. :-/

So... can you convince me that this will fail in a deterministic, sane way with no possibility of memory corruption bugs?

File: Firestore/core/src/firebase/firestore/model/delete_mutation.cc [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+static_assert(
+    sizeof(Mutation) == sizeof(DeleteMutation),
+    "DeleteMutation may not have additional members (everything goes in Rep)");
+
+DeleteMutation::DeleteMutation(DocumentKey key, Precondition precondition)
+    : Mutation(std::make_shared<Rep>(std::move(key), std::move(precondition))) {
+}
+
+DeleteMutation::DeleteMutation(const Mutation& mutation) : Mutation(mutation) {
+  HARD_ASSERT(type() == Type::Delete);
```
Comment (mikelehen) [inline Firestore/core/src/firebase/firestore/model/delete_mutation.cc]: FWIW- It occurs to me that this means casting an invalid instance will result in a nullptr dereference.  Is that OK?

File: Firestore/core/src/firebase/firestore/model/mutation.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/model/mutation.h]: These posts are both problematic because they intentionally steer toward what the language guarantees, and the language gives itself a huge out with the notion of undefined behavior. While it's true that the standard doesn't specify what happens, there are no practically useful systems in the wild where dereferencing null doesn't crash somehow. On UNIX systems including the iPhone, the process will receive signal 11 (SEGV) and that's the end of it.

Admittedly this may not result in a stack trace. Under Xcode it certainly pauses the process exactly at the point of failure, but this isn't going to happen for end users. A hard assert is slight better in that it will result in a stack.

Note that the code this is replacing wasn't any better. Using moved-from `std::shared_ptr<Mutation>` is equivalently hazardous, though perhaps by making the pointer more obvious invites more caution. IDK.

I'm willing to insert the checks if we can do so without littering it all over the code. The downside is that if it's in a central location it may be harder to pinpoint what's going on.

Anyway, to sum up:
  * Users will definitely crash on a null pointer dereference
  * It's not likely to be all that useful a message
  * Centralizing the check would make it not too cumbersome
  * I worry that the result may not end up all that useful either.

[the remark this answers is not in the document]
Comment (mikelehen) [reply Firestore/core/src/firebase/firestore/model/mutation.h]: I think ensuring we get a stack trace is probably worthwhile.  To centralize we could probably put the check in the `rep()` getter and then change all the implementation methods to go through the getter.

File: Firestore/core/src/firebase/firestore/model/delete_mutation.cc (hunk 8 above)
Comment (mikelehen) [reply Firestore/core/src/firebase/firestore/model/delete_mutation.cc]: If we put the check in `rep()` and make sure `type()` uses the `rep()` getter, then that would work.

File: Firestore/core/src/firebase/firestore/model/mutation.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/model/mutation.h]: I've reimplemented `rep()` as `*NOT_NULL(rep_)`.

Also since we're going to be null checking a lot more with this, I made null checks cheaper (by ~30%).
